controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -369,44 +375,58 @@ public void initializeStreamWriters(final EventStreamWriter<CommitEvent> commitW
-
-                    // Step 3. Create txn node in the store.
-                    CompletableFuture<VersionedTransactionData> txnFuture = createTxnInStore(scope, stream, lease,
-                            ctx, maxExecutionPeriod, txnId, addIndex, ctx.getRequestId());
-
-                    // Step 4. Notify segment stores about new txn.
-                    CompletableFuture<List<StreamSegmentRecord>> segmentsFuture = txnFuture.thenComposeAsync(txnData ->
-                            streamMetadataStore.getSegmentsInEpoch(scope, stream, txnData.getEpoch(), ctx, executor), executor);
-
-                    CompletableFuture<Void> notify = segmentsFuture.thenComposeAsync(activeSegments ->
-                            notifyTxnCreation(scope, stream, activeSegments, txnId, ctx.getRequestId()), executor)
-                                                                   .whenComplete((v, e) ->
-                            // Method notifyTxnCreation ensures that notification completes
-                            // even in the presence of n/w or segment store failures.
-                            log.trace(ctx.getRequestId(), "Txn={}, notified segments stores", txnId));
-
-                    // Step 5. Start tracking txn in timeout service
-                    return notify.whenCompleteAsync((result, ex) -> {
-                        addTxnToTimeoutService(scope, stream, lease, maxExecutionPeriod, txnId, txnFuture, ctx.getRequestId());
-                    }, executor).thenApplyAsync(v -> {
-                        List<StreamSegmentRecord> segments = segmentsFuture.join().stream().map(x -> {
-                            long generalizedSegmentId = RecordHelper.generalizedSegmentId(x.segmentId(), txnId);
-                            int epoch = NameUtils.getEpoch(generalizedSegmentId);
-                            int segmentNumber = NameUtils.getSegmentNumber(generalizedSegmentId);
-                            return StreamSegmentRecord.builder().creationEpoch(epoch).segmentNumber(segmentNumber)
-                                    .creationTime(x.getCreationTime()).keyStart(x.getKeyStart()).keyEnd(x.getKeyEnd()).build();
-                        }).collect(Collectors.toList());
-
-                        return new ImmutablePair<>(txnFuture.join(), segments);
-                    }, executor);
-                }), e -> {
-            Throwable unwrap = Exceptions.unwrap(e);
-            return unwrap instanceof StoreException.WriteConflictException || unwrap instanceof StoreException.DataNotFoundException;
-        }, 5, executor));
+        return validate.thenCompose(validated -> checkTxnRateLimit(scope, stream, ctx))
+                .thenCompose(limitCheckPassed -> {
+                if (!limitCheckPassed) {
+                    throw new RuntimeException("Create transaction failed because Open Transactions limit exceeded. Try again later.");
```

Comment:
should we add delays rather than throw this back to the caller?

Reply:
For a POC I think this should be good